### PR TITLE
Find correct pid when pgrep returns multiple pids

### LINKED ISSUE #22: Pgrep qemu-system is not robust
Currently, gdb-pt-dump parses the output of `pgrep qemu-system` which is not robust if there are multiple PIDs returned.

Is there any better we can do here?

Relevant issue: https://github.com/pwndbg/pwndbg/issues/1587#issuecomment-1438379102

### PR DESCRIPTION
**Pull request by PandaWill**
Fixes #22

I.e. parsing:
```
(gdb) monitor info cpus
* CPU #0: thread_id=1234
  CPU #1: thread_id=1235
```

Note that this is a thread id. We could parse proc to find the process id but it's not necessary since threads share memory. So /proc/[tid]/mem is the same as /proc/[pid]/mem.

### CODE REVIEW

**Comment by PandaWill** [comment]
Note this fixes pwndbg/pwndbg#1587

**Comment by PandaWill** [comment]
Hi @martinradev 

Please could you review/merge this so that we can get the pwndbg bug fixed?

Thanks.

File: pt.py [hunk 1]
```diff
@@ -157,9 +158,14 @@ def __init__(self):
...
         # Get quick access to physical memory.
-        proc = os.popen("pgrep qemu-system")
-        self.pid = int(proc.read().strip(), 10)
-        proc.close()
+
+        # Technically `info cpus` gives us a thread id, but this is sufficient for use with /proc/[tid]/mem
+        cpus = gdb.execute("monitor info cpus", to_string = True)
+        m = re.search('thread_id=(\d+)', cpus)
+        if not m:
+            raise Exception(f"Could not find qemu pid")
```
Comment (martinradev) [inline pt.py:166]: Could you please add a fallback option to use pgrep?

The problems I see with this approach:
- qemu and gdb may be in different pid namespaces
- the "monitor info cpus" may return different output or may not be available across different qemu versions.

Something like this:
1) Try to  get info via "monitor info cpus". If success, check that process actually exists and it's qemu
2) If fails, print message and print  info via pgrep qemu-system.

Would this make sense?

**Comment by PandaWill** [comment]
I hadn't even considered namespaces, thanks for pointing that out!

I've come up with a more robust method now that is namespace agnostic.

We can add a chardev file backend via the monitor commands. If we create this with a known file path (e.g /tmp/foo123), we can then check for that path as a link target in /proc/[pid]/fd.

This even works for mount namespaces. It won't be the same file, but the path relative to the root of the mount namespace will be.

We only do this when pgrep returns more than one pid.

**Comment by martinradev** [comment]
Looks reasonable. I should really add a proper integration test suite for this project.

Thanks a lot Will!

**Pull request merged**